source/windows/secure_channel_tls_handler.c

```diff
@@ -1283,15 +1310,23 @@ static int s_process_read_message(
                             break;
                         }
                     }
-                    /* prevent a deadlock due to downstream handlers wanting more data, but we have an incomplete
-                       record, and the amount they're requesting is less than the size of a tls record. */
-                    size_t window_size = slot->window_size;
-                    if (!window_size &&
-                        aws_channel_slot_increment_read_window(slot, sc_handler->estimated_incomplete_size)) {
-                        err = AWS_OP_ERR;
-                    } else {
-                        sc_handler->estimated_incomplete_size = 0;
-                        err = AWS_OP_SUCCESS;
+
+                    size_t downstream_window = SIZE_MAX;
+
+                    if (sc_handler->slot->adj_right) {
+                        downstream_window = aws_channel_slot_downstream_read_window(sc_handler->slot);
+                    }
+                    if (downstream_window > 0) {
+                        /* prevent a deadlock due to downstream handlers wanting more data, but we have an incomplete
+                           record, and the amount they're requesting is less than the size of a tls record. */
+                        size_t window_size = slot->window_size;
+                        if (!window_size &&
+                            aws_channel_slot_increment_read_window(slot, sc_handler->estimated_incomplete_size)) {
+                            err = AWS_OP_ERR;
+                        } else {
+                            sc_handler->estimated_incomplete_size = 0;
+                            err = AWS_OP_SUCCESS;

```

Comment:
by moving this line under a new `if (downstream_window > 0)` branch, the `err` doesn't get reset to 0

OMG THIS LOOP IS BRAIN MELTING